Compile errors using Typescript 2.2.1

Installed the latest typescript build (2.2.1) and tried building antlr4ts w/ `npm install`.   During the **tsc** step it generated errors that seem reasonably related to the new features in TypeScript 2.2.    I'll dig into them.

Result:
```
src/Parser.ts(81,59): error TS2344: Type 'string' does not satisfy the constraint 'object'.
src/Parser.ts(810,4): error TS2322: Type 'RuleContext | undefined' is not assignable to type 'RuleContext'.
  Type 'undefined' is not assignable to type 'RuleContext'.
src/dfa/DFAState.ts(173,56): error TS2345: Argument of type 'DFAState | undefined' is not assignable to parameter of type 'DFAState'.
  Type 'undefined' is not assignable to type 'DFAState'.
src/dfa/DFAState.ts(179,53): error TS2345: Argument of type 'DFAState | undefined' is not assignable to parameter of type 'DFAState'.
  Type 'undefined' is not assignable to type 'DFAState'.
```

--- a/src/Parser.ts
+++ b/src/Parser.ts
@@ -799,7 +799,8 @@ export abstract class Parser extends Recognizer<Token, ParserATNSimulator> {
 	 *  This is very useful for error messages.
 	 */
 
-	getRuleInvocationStack(p: RuleContext | undefined = this._ctx): string[] {
+	getRuleInvocationStack(ctx: RuleContext = this._ctx): string[] {
+		let p: RuleContext | undefined = ctx;  		// Workaround for Microsoft/TypeScript#14487
 		let ruleNames: string[] = this.ruleNames;
 		let stack: string[] = [];
 		while (p != null) {

--- a/src/Parser.ts
+++ b/src/Parser.ts
@@ -807,7 +807,7 @@ export abstract class Parser extends Recognizer<Token, ParserATNSimulator> {
 			let ruleIndex: number = p.ruleIndex;
 			if (ruleIndex < 0) stack.push("n/a");
 			else stack.push(ruleNames[ruleIndex]);
-			p = p._parent;
+			p = p._parent as RuleContext;
 		}
 		return stack;
 	}

--- a/src/Parser.ts
+++ b/src/Parser.ts
@@ -78,7 +78,7 @@ export abstract class Parser extends Recognizer<Token, ParserATNSimulator> {
 	 *
 	 * @see ATNDeserializationOptions.isGenerateRuleBypassTransitions
 	 */
-	private static readonly bypassAltsAtnCache = new WeakMap<string, ATN>();
+	private static readonly bypassAltsAtnCache = new Map<string, ATN>();
 
 	/**
 	 * The error handling strategy for the parser. The default value is a new

--- a/src/dfa/DFAState.ts
+++ b/src/dfa/DFAState.ts
@@ -170,13 +170,13 @@ export class DFAState {
 		if (map.has(-1)) {
 			if (map.size === 1) {
 				let result = new Map<number, DFAState>();
-				result.set(PredictionContext.EMPTY_FULL_STATE_KEY, map.get(-1));
+				result.set(PredictionContext.EMPTY_FULL_STATE_KEY, map.get(-1) as DFAState);
 				return result;
 			}
 			else {
 				let removed = map.get(-1);
 				map.delete(-1);
-				map.set(PredictionContext.EMPTY_FULL_STATE_KEY, removed);
+				map.set(PredictionContext.EMPTY_FULL_STATE_KEY, removed as DFAState);
 			}
 		}
 

--- a/src/dfa/DFAState.ts
+++ b/src/dfa/DFAState.ts
@@ -167,16 +167,16 @@ export class DFAState {
 
 	getContextEdgeMap(): Map<number, DFAState> {
 		let map = new Map<number, DFAState>(this.contextEdges);
-		if (map.has(-1)) {
+		let existing = map.get(-1);
+		if (existing !== undefined) {
 			if (map.size === 1) {
 				let result = new Map<number, DFAState>();
-				result.set(PredictionContext.EMPTY_FULL_STATE_KEY, map.get(-1) as DFAState);
+				result.set(PredictionContext.EMPTY_FULL_STATE_KEY, existing);
 				return result;
 			}
 			else {
-				let removed = map.get(-1);
 				map.delete(-1);
-				map.set(PredictionContext.EMPTY_FULL_STATE_KEY, removed as DFAState);
+				map.set(PredictionContext.EMPTY_FULL_STATE_KEY, existing);
 			}
 		}
 

